frame/support/src/traits/try_runtime.rs

```diff
@@ -0,0 +1,172 @@
+	/// to.
+	fn set_temp_storage<T: codec::Encode>(data: T, at: &str) {
+		sp_io::storage::set(&Self::storage_key(at), &data.encode());
+	}
+}
+
+impl<U: OnRuntimeUpgrade> OnRuntimeUpgradeHelpersExt for U {}
+
+// Which state tests to execute.
+#[derive(codec::Encode, codec::Decode, Clone)]
+pub enum Select {
+	/// None of them.
+	None,
+	/// All of them.
+	All,
+	/// Run a fixed number of them in a round robin manner.
+	RoundRobin(u32),
+	/// Run only pallets who's name matches the given list.
+	///
+	/// Pallet names are obtained from [`PalletInfoAccess`].
+	Only(Vec<Vec<u8>>),
+}
+
+impl Default for Select {
+	fn default() -> Self {
+		Select::None
+	}
+}
+
+impl sp_std::fmt::Debug for Select {
+	fn fmt(&self, f: &mut sp_std::fmt::Formatter<'_>) -> sp_std::fmt::Result {
+		match self {
+			Select::RoundRobin(x) => write!(f, "RoundRobin({})", x),
+			Select::Only(x) => write!(
+				f,
+				"Only({:?})",
+				x.iter().map(|x| sp_std::str::from_utf8(x).unwrap()).collect::<Vec<_>>(),
+			),
+			Select::All => write!(f, "All"),
+			Select::None => write!(f, "None"),
+		}
+	}
+}
+
+#[cfg(feature = "std")]
+impl sp_std::str::FromStr for Select {
+	type Err = &'static str;
+	fn from_str(s: &str) -> Result<Self, Self::Err> {
+		match s {
+			"all" | "All" => Ok(Select::All),
+			"none" | "None" => Ok(Select::None),
+			_ =>
+				if s.starts_with("rr-") {
+					let count = s
+						.split_once('-')
+						.and_then(|(_, count)| count.parse::<u32>().ok())
+						.ok_or("failed to parse count")?;
+					Ok(Select::RoundRobin(count))
+				} else {
+					let pallets = s.split(',').map(|x| x.as_bytes().to_vec()).collect::<Vec<_>>();
+					Ok(Select::Only(pallets))
+				},
+		}
+	}
+}
+
+/// Execute some checks a pallet to ensure the state is consistent.
+pub trait TryState<BlockNumber> {
+	/// Execute the state sanity checks.
+	fn try_state(_: BlockNumber, _: Select) -> Result<(), &'static str>;
+}
+
+#[cfg_attr(
+	all(feature = "try-runtime", not(feature = "tuples-96"), not(feature = "tuples-128")),
+	impl_for_tuples(64)
+)]
+#[cfg_attr(
+	all(feature = "try-runtime", feature = "tuples-96", not(feature = "tuples-128")),
+	impl_for_tuples(96)
+)]
+#[cfg_attr(all(feature = "try-runtime", feature = "tuples-128"), impl_for_tuples(128))]
+impl<BlockNumber: Clone + AtLeast32BitUnsigned> TryState<BlockNumber> for Tuple {
+	for_tuples!( where #( Tuple: crate::traits::PalletInfoAccess )* );
+	fn try_state(n: BlockNumber, targets: Select) -> Result<(), &'static str> {
+		match targets {
+			Select::None => Ok(()),
+			Select::All => {
+				let mut result = Ok(());
+				for_tuples!( #( result = result.and(Tuple::try_state(n.clone(), targets.clone())); )* );
+				result
+			},
+			Select::RoundRobin(len) => {
+				let functions: &[fn(BlockNumber, Select) -> Result<(), &'static str>] =
+					&[for_tuples!(#( Tuple::try_state ),*)];
+				let skip = n.clone() % (len as u32).into();
+				let skip: u32 = skip.try_into().ok().expect("TODO");
```

Comment:
Maybe easier to read this way?
```suggestion
let skip: u32 = n.clone().try_into().ok().expect("Block numbers can be cast to a u32; qed");
let skip = skip % len;
```